include/aos/common/resourcemonitor.hpp

```diff
@@ -0,0 +1,301 @@
+};
+
+/**
+ * Instance resource monitor parameters.
+ */
+struct InstanceMonitorParams {
+    InstanceIdent                                 mInstanceIdent;
+    StaticArray<PartitionInfo, cMaxNumPartitions> mPartitions;
+};
+
+/**
+ * Resource usage provider interface.
+ */
+class ResourceUsageProviderItf {
+public:
+    /**
+     * Destructor
+     */
+    virtual ~ResourceUsageProviderItf() = default;
+
+    /**
+     * Initializes resource usage provider
+     *
+     * @return Error
+     */
+    virtual Error Init() = 0;
+
+    /**
+     * Returns node info
+     *
+     * @param systemInfo system info
+     * @return Error
+     */
+    virtual Error GetNodeInfo(NodeInfo& systemInfo) const = 0;
+
+    /**
+     * Returns resource config
+     *
+     * @param nodeID node ident
+     * @param monitoringData monitoring data
+     * @return Error
+     */
+    virtual Error GetNodeMonitoringData(const String& nodeID, MonitoringData& monitoringData) = 0;
+
+    /**
+     * Gatherings instance monitoring data
+     *
+     * @param instance instance ident
+     * @param monitoringData monitoring data
+     * @return Error
+     */
+    virtual Error GetInstanceMonitoringData(const String& instanceID, MonitoringData& monitoringData) = 0;
+};
+
+/**
+ * Monitor sender interface.
+ */
+class SenderItf {
+public:
+    /**
+     * Sends monitoring data
+     *
+     * @param monitoringData monitoring data
+     * @return Error
+     */
+    virtual Error SendMonitoringData(const NodeMonitoringData& monitoringData) = 0;
+};
+
+/**
+ * Resource monitor interface.
+ */
+class ResourceMonitorItf {
+public:
+    /**
+     * Destructor
+     */
+    virtual ~ResourceMonitorItf() = default;
+
+    /**
+     * Gets the node info object
+     *
+     * @param nodeInfo node info
+     * @return Error
+     */
+    virtual Error GetNodeInfo(NodeInfo& nodeInfo) const = 0;
+
+    /**
+     * Starts instance monitoring
+     *
+     * @param instanceID instance ident
+     * @param monitoringConfig monitoring config
+     * @return Error
+     */
+    virtual Error StartInstanceMonitoring(const String& instanceID, const InstanceMonitorParams& monitoringConfig) = 0;
+
+    /**
+     * Stops instance monitoring
+     *
+     * @param instanceID instance ident
+     * @return Error
+     */
+    virtual Error StopInstanceMonitoring(const String& instanceID) = 0;
+};
+
+/**
+ * Resource monitor.
+ */
+class ResourceMonitor : public ResourceMonitorItf, public SubscriberItf {
+public:
+    /**
+     * Constructor
+     */
+    ResourceMonitor() = default;
+
+    /**
+     * Destructor
+     */
+    ~ResourceMonitor();
+
+    /**
+     * Initializes resource monitor
+     *
+     * @param resourceUsageProvider resource usage provider
+     * @param monitorSender monitor sender
+     * @return Error
+     */
+    Error Init(
+        ResourceUsageProviderItf& resourceUsageProvider, SenderItf& monitorSender, aos::SubscribableItf& subscribable);
```

Comment:
Remove subscription itf from this commit.

Reply:
Done